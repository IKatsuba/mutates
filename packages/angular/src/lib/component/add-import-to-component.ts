import type { ClassDeclaration } from 'ts-morph';

import { pushToDecoratorArrayProperty } from '../helpers/push-to-decorator-array-property';

export function addImportToComponent(
  classDeclaration: ClassDeclaration,
  importName: string,
  { unique = false }: { unique?: boolean } = {},
): void {
  pushToDecoratorArrayProperty(classDeclaration, 'Component', 'imports', importName, {
    unique,
    forceToArray: true,
  });
}
