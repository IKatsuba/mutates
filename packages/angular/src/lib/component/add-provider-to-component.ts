import type { ClassDeclaration } from 'ts-morph';

import { pushToDecoratorArrayProperty } from '../helpers/push-to-decorator-array-property';

export function addProviderToComponent(
  classDeclaration: ClassDeclaration,
  provider: string,
  { unique = false }: { unique?: boolean } = {},
): void {
  pushToDecoratorArrayProperty(classDeclaration, 'Component', 'providers', provider, {
    unique,
    forceToArray: true,
  });
}
