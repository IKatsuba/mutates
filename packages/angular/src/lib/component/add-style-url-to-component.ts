import type { ClassDeclaration } from 'ts-morph';

import { pushToDecoratorArrayProperty } from '../helpers/push-to-decorator-array-property';

export function addStyleUrlToComponent(
  classDeclaration: ClassDeclaration,
  styleUrl: string,
  { unique = false }: { unique?: boolean } = {},
): void {
  pushToDecoratorArrayProperty(classDeclaration, 'Component', 'styleUrls', styleUrl, {
    unique,
  });
}
