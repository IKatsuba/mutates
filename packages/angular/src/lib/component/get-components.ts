import type { ClassDeclaration } from 'ts-morph';

import { getClasses, Pattern, Query, StructureType } from '@mutates/core';

export function getComponents(
  pattern: Pattern,
  query?: Query<Omit<StructureType<ClassDeclaration>, 'kind'>>,
): ClassDeclaration[] {
  return getClasses(pattern, query).filter(isComponent);
}

export function isComponent(declaration: ClassDeclaration): boolean {
  return !!declaration.getDecorator('Component');
}
