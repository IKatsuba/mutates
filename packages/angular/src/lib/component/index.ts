export * from './add-import-to-component';
export * from './add-provider-to-component';
export * from './add-style-url-to-component';
