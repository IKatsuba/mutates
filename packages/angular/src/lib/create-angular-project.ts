import { Tree } from '@angular-devkit/schematics';
import { ProjectOptions } from 'ts-morph';

import { createProject } from '@mutates/core';

import { NgTreeFileSystem } from './ng-tree-file-system';

export function createAngularProject(tree: Tree, options?: Omit<ProjectOptions, 'fileSystem'>) {
  return createProject(new NgTreeFileSystem(tree), options);
}
