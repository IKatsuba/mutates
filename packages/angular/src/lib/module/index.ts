export * from './add-bootstrap-to-ng-module';
export * from './add-declaration-to-ng-module';
export * from './add-entry-component-to-ng-module';
export * from './add-export-to-ng-module';
export * from './add-import-to-ng-module';
export * from './add-provider-to-ng-module';
export * from './add-schema-to-ng-module';
