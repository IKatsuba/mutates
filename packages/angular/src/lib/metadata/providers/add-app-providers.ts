import { ArrayLiteralExpression, ObjectLiteralExpression, SyntaxKind } from 'ts-morph';

import { getObjectProperties, pushToArrayIfNotExists } from '@mutates/core';

import { getBootstrapApplicationFn } from '../../bootstrap';
import { getAppModule } from '../../module';
import { getMetadataProperty } from '../index';

export function getAppProviders(mainFilePath: string): ArrayLiteralExpression | undefined {
  const appModule = getAppModule(mainFilePath);

  if (appModule) {
    return getMetadataProperty(appModule, 'providers')?.getInitializerIfKind(
      SyntaxKind.ArrayLiteralExpression,
    );
  }

  const boostrapApplicationFn = getBootstrapApplicationFn(mainFilePath);

  if (!boostrapApplicationFn) {
    return;
  }
  const config = (boostrapApplicationFn.getArguments().at(1) ??
    boostrapApplicationFn.addArgument(`{providers: []}`)) as ObjectLiteralExpression;

  return getObjectProperties(config, { name: 'providers' })
    .at(0)
    ?.getInitializerIfKind(SyntaxKind.ArrayLiteralExpression);
}

export function addAppProviders(mainFilePath: string, providers: string[]): void {
  const appProviders = getAppProviders(mainFilePath);

  if (!appProviders) {
    return;
  }

  pushToArrayIfNotExists(appProviders, ...providers);
}
