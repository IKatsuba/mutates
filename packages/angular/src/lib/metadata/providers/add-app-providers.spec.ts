import { Tree } from '@angular-devkit/schematics';

import { createSourceFile, readFileSync, saveProject } from '@mutates/core';

import { createAngularProject } from '../../create-angular-project';
import { createTestingTree } from '../../testing';
import { addAppProviders } from './add-app-providers';

describe('addAppProviders', () => {
  let host: Tree;
  beforeEach(() => {
    host = createTestingTree();

    createAngularProject(host);
  });

  it('should add providers to the app module', () => {
    createSourceFile(
      'src/main.ts',
      `import {platformBrowserDynamic} from '@angular/platform-browser-dynamic';
import {AppModule} from './app/app.module';
import {environment} from './environments/environment';

if (environment.production) {
  enableProdMode();
}

platformBrowserDynamic()
  .bootstrapModule(AppModule)
  .catch(err => console.log(err));
`,
    );

    createSourceFile(
      'src/app/app.module.ts',
      `import {NgModule} from '@angular/core';
    
@NgModule({
  providers: [],
})
export class AppModule {}
`,
    );

    addAppProviders('src/main.ts', ['AppService']);

    saveProject();

    expect(readFileSync('src/app/app.module.ts')).matchSnapshot();
  });

  it('should add providers to the bootstrapApplication function', () => {
    createSourceFile(
      'src/main.ts',
      `import {bootstrapApplication} from '@angular/platform-browser';
import {AppComponent} from './app/app.component';
import {environment} from './environments/environment';

if (environment.production) {
  enableProdMode();
}

bootstrapApplication(AppComponent)
`,
    );

    addAppProviders('src/main.ts', ['AppService']);

    saveProject();

    expect(readFileSync('src/main.ts')).matchSnapshot();
  });
});
