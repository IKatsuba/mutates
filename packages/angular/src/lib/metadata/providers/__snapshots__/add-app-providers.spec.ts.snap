// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`addAppProviders > should add providers to the app module 1`] = `
"import {NgModule} from '@angular/core';
    
@NgModule({
  providers: [AppService],
})
export class AppModule {}
"
`;

exports[`addAppProviders > should add providers to the bootstrapApplication function 1`] = `
"import {bootstrapApplication} from '@angular/platform-browser';
import {AppComponent} from './app/app.component';
import {environment} from './environments/environment';

if (environment.production) {
  enableProdMode();
}

bootstrapApplication(AppComponent, {providers: [AppService]})
"
`;
